### PR TITLE
Zimperium MTD Threats - add connector

### PR DESCRIPTION
**Pull request by lvoloshyn-sekoia**
https://github.com/SekoiaLab/detection-rules/issues/3926

## Summary by Sourcery

Introduce a new Zimperium module that collects Mobile Threat Defence threats and forwards them to SEKOIA.IO, including authentication, metrics, and account validation support.

New Features:
- Add a Mobile Threat Defence connector to fetch threats from Zimperium MTD and forward them to an intake on a scheduled basis.
- Introduce Zimperium API authentication and client utilities with token management and rate limiting.
- Provide an account validator for Zimperium credentials within the automation framework.

Enhancements:
- Track Zimperium collection and forwarding activity with Prometheus metrics for incoming messages, forwarded events, lag, and processing duration.
- Persist connector cursor and a deduplication cache to avoid reprocessing already seen events.
- Add structured logging setup for the Zimperium module and wire it into a runnable main entrypoint.

Build:
- Add a dedicated Zimperium pyproject configuration and Dockerfile to build and run the module in the automation environment.

Deployment:
- Provide connector configuration JSON stubs for the MTD threats trigger/connector.

Documentation:
- Add an initial changelog file for the new Zimperium module.

Tests:
- Add tests for the Mobile Threat Defence connector batching logic, including handling of empty and non-empty threat pages and scheduling delays.

### CODE REVIEW

**Review by LenaigKaliou — APPROVED** [review]
LGTM

File: Zimperium/trigger_mtd_threats.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      },
+      "intake_key": {
+        "description": "Intake key to use when sending events",
+        "type": "string"
+      },
+      "chunk_size": {
+        "type": "integer",
+        "description": "The max size of chunks for the batch processing",
+        "minimum": 1,
+        "maximum": 1000,
```
Comment (squioc) [inline Zimperium/trigger_mtd_threats.json]: [REQUIRED] The connector allows 2000 as maximum. Can you sync both manifest files? Thank you

File: Zimperium/mise.toml
Comment (squioc) [inline Zimperium/mise.toml:1]: [REQUIRED] My bad!! The expected name is `mise.toml` (not `mie.toml` like written in he migration guide 😒 ). Can you fix it?

File: Zimperium/CHANGELOG.md
Comment (squioc) [inline Zimperium/CHANGELOG.md:1]: [REQUIRED] Can you add a first entry talking about the initial version? Thank you.

File: Zimperium/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,57 @@
+[project]
+name = "zimperium-automation-module"
+version = "0.0"
+description = ""
+authors = []
+requires-python = ">=3.11,<3.15"
+
+dependencies = [
+    "sekoia-automation-sdk>=1.23.0",
+]
```
Comment (squioc) [inline Zimperium/pyproject.toml:17]: [REQUIRED] Can you add requests and structlog as dependencies? I know that the sdk depends on requests, but, it's preferable having a direct dependencies here to show the link. thank you.

File: Zimperium/zimperium_modules/account_validator.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+class ZimperiumAccountValidator(AccountValidator):
+    module: ZimperiumModule
+
+    @cached_property
+    def client(self) -> ApiClient:
+        return ApiClient(
+            base_url=self.module.configuration.base_url,
+            client_id=self.module.configuration.client_id,
+            client_secret=self.module.configuration.client_secret,
+        )
```
Comment (squioc) [inline Zimperium/zimperium_modules/account_validator.py]: [RECOMMENDED] Is it required to define this method? I don't see usage of it here.

File: Zimperium/zimperium_modules/connector_threats.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
+import time
+from datetime import datetime, timedelta, timezone
+from functools import cached_property
+from posixpath import join as urljoin
+from typing import Any, Generator
+
+import orjson
+import requests
+from cachetools import Cache, LRUCache
+from dateutil.parser import isoparse
```
Comment (squioc) [inline Zimperium/zimperium_modules/connector_threats.py:11]: [REQUIRED] It seems to be unused

```suggestion
```

File: Zimperium/zimperium_modules/connector_threats.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        super().__init__(*args, **kwargs)
+
+        self.context = PersistentJSON("context.json", self.data_path)
+        self.cursor = CheckpointDatetime(
+            path=self.data_path,
+            start_at=timedelta(days=200),
+            ignore_older_than=timedelta(days=300),
+        )
+
+        self.cache_size = 2000
```
Comment (squioc) [inline Zimperium/zimperium_modules/connector_threats.py]: [RECOMMENDED] Can we set it from env? like [salesforce](https://github.com/SEKOIA-IO/automation-library/blob/develop/Salesforce/salesforce/connector.py#L70)

File: Zimperium/zimperium_modules/client/auth.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                per_second=ratelimit_per_second,
+                max_retries=Retry(
+                    total=5,
+                    backoff_factor=1,
+                ),
+            ),
+        )
+
+    def get_credentials(self) -> ZimperiumApiCredentials:
+        current_dt = datetime.utcnow()
```
Comment (squioc) [inline Zimperium/zimperium_modules/client/auth.py]: [RECOMMENDED] Deprecated in newer Python version
Please use:

```suggestion
        current_dt = datetime.now(datetime.UTC)
```

File: Zimperium/zimperium_modules/client/auth.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    AUTH_TTL = 60 * 60
+
+    def __init__(
+        self,
+        base_url: str,
+        client_id: str,
+        client_secret: str,
+        ratelimit_per_second: int = 10,
+    ) -> None:
+        self.__authorization_url = urljoin(base_url, "api/auth/v1/api_keys/login")
```
Comment (squioc) [inline Zimperium/zimperium_modules/client/auth.py]: [RECOMMENDED] It seems that this property is not used. It redefined L64.
According to the logic in the `get_credentials` method, I propose deleting this property:

```suggestion
```

File: Zimperium/manifest.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        "title": "Client Secret",
+        "type": "string"
+      }
+    },
+    "required": [
+      "base_url",
+      "client_id",
+      "client_secret"
+    ],
+    "secrets": "client_secret"
```
Comment (squioc) [inline Zimperium/manifest.json]: [REQUIRED] `secrets` must be an array:

```suggestion
    "secrets": ["client_secret"]
```

File: Zimperium/zimperium_modules/logging.py
Comment (squioc) [inline Zimperium/zimperium_modules/logging.py:1]: [RECOMMENDED] This module is not used in the connector. Do you keep it or do you use it in `connector_threats.py`?

File: Zimperium/zimperium_modules/connector_threats.py
Comment (squioc) [inline Zimperium/zimperium_modules/connector_threats.py:1]: [SUGGESTION] Can we use a [timestepper](https://github.com/SEKOIA-IO/sekoia-automation-sdk/blob/main/sekoia_automation/helpers/timestepper.py) to define the period of time to fetch?

**Review by squioc — CHANGES_REQUESTED** [review]
Hello,
Thank you for the work.
Some comments were made.

File: Zimperium/manifest.json [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      "client_secret"
+    ],
+    "secrets": ["client_secret"]
+  },
+  "description": "Zimperium is a global leader in mobile security, providing an AI-driven platform designed to protect enterprise devices and applications from advanced cyber threats. Their technology offers real-time, on-device protection against phishing, malware, and network attacks across iOS, Android, and ChromeOS. By integrating security into the entire app lifecycle, they help organizations secure critical data and maintain regulatory compliance in a mobile-first world.",
+  "name": "Zimperium",
+  "slug": "zimperium",
+  "categories": [],
+  "uuid": "0f7801d0-6fc6-4583-beff-05d5ab1dda20",
+  "version": "0.1.0"
```
Comment (squioc) [inline Zimperium/manifest.json]: [REQUIRED] To match the CHANGELOG.md
```suggestion
  "version": "1.0.0"
```

File: Zimperium/zimperium_modules/connector_threats.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        params: dict[str, str | int] = {
+            "module": "ZIPS",
+            "after": from_date.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
+            "before": to_date.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
+            "page": page_num,
+            "size": self.configuration.chunk_size,
+            "sort": "timestamp,asc",
+        }
+
+        while True:
```
Comment (squioc) [inline Zimperium/zimperium_modules/connector_threats.py]: [RECOMMENDED]
```suggestion
        while self.running:
```

File: Zimperium/connector_mtd_threats.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  "uuid": "378c5c4e-28ce-57b8-b287-1762a4c38c80",
+  "docker_parameters": "MobileThreatDefenceConnector",
+  "arguments": {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "properties": {
+      "frequency": {
+        "type": "integer",
+        "description": "Batch frequency in seconds",
+        "default": 60
+      },
```
Comment (squioc) [inline Zimperium/connector_mtd_threats.json:13]: [REQUIRED] These parameters are missing from the configuration:
```suggestion
      },
      "timedelta": {
        "type": "integer",
        "description": "The temporal shift, in the past, in minutes, the connector applies when fetching the events (default to 5 minutes ago)",
        "default": 5
      },
      "start_time": {
        "type": "integer",
        "description": "The number of hours from which events should be queried.",
        "default": 1
      },
```

File: Zimperium/trigger_mtd_threats.json [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  "uuid": "378c5c4e-28ce-57b8-b287-1762a4c38c80",
+  "docker_parameters": "MobileThreatDefenceConnector",
+  "arguments": {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "properties": {
+      "frequency": {
+        "type": "integer",
+        "description": "Batch frequency in seconds",
+        "default": 60
+      },
```
Comment (squioc) [inline Zimperium/trigger_mtd_threats.json:13]: [REQUIRED] These parameters are missing from the configuration:
```suggestion
      },
      "timedelta": {
        "type": "integer",
        "description": "The temporal shift, in the past, in minutes, the connector applies when fetching the events (default to 5 minutes ago)",
        "default": 5
      },
      "start_time": {
        "type": "integer",
        "description": "The number of hours from which events should be queried.",
        "default": 1
      },
```

File: Zimperium/zimperium_modules/connector_threats.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+logger = get_logger()
+
+
+class MobileThreatDefenceConnectorConfiguration(DefaultConnectorConfiguration):
+    chunk_size: int = 1000
+
+    frequency: int = 60
+    timedelta: int = 5
+    start_time: int = 10
```
Comment (squioc) [inline Zimperium/zimperium_modules/connector_threats.py]: [SUGGESTION] Are you sure we want to start 10 hours in the past at the first run?

**Review by squioc — CHANGES_REQUESTED** [review]
Thank you for the changes.
Some new comments were added.

**Review by squioc — APPROVED** [review]
LGTM. Thank you for the work
